src/spv/spv_wrapper.cpp

```diff
@@ -662,13 +700,218 @@ UniValue CSpvWrapper::ListTransactions()
     return result;
 }
 
+UniValue CSpvWrapper::GetHTLCReceived(const std::string& addr)
+{
+    if (!addr.empty() && !BRAddressIsValid(addr.c_str()))
+    {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid address");
+    }
+
+    UInt160 addressFilter = UINT160_ZERO;
+    if (!addr.empty())
+    {
+        BRAddressHash160(&addressFilter, addr.c_str());
+    }
+
+    auto htlcTransactions = BRListHTLCReceived(wallet, addressFilter);
+
+    UniValue result(UniValue::VARR);
+    for (const auto& txInfo : htlcTransactions)
+    {
+        uint256 txid = to_uint256(txInfo.first->txHash);
+        uint64_t confirmations{0};
+        uint32_t blockHeight = ReadTxBlockHeight(txid);
+
+        if (blockHeight != std::numeric_limits<int32_t>::max())
+        {
+            confirmations = spv::pspv->GetLastBlockHeight() - blockHeight + 1;
+        }
+
+        UniValue item(UniValue::VOBJ);
+        uint64_t output = txInfo.second;
+        item.pushKV("txid", txid.ToString());
+        item.pushKV("vout", output);
+        item.pushKV("amount", ValueFromAmount(txInfo.first->outputs[output].amount));
+        item.pushKV("address", txInfo.first->outputs[output].address);
+        item.pushKV("confirms", confirmations);
+
+        uint256 spent;
+        if (BRWalletTxSpent(wallet, txInfo.first, output, spent))
+        {
+            blockHeight = ReadTxBlockHeight(spent);
+            confirmations = 0;
+            if (blockHeight != std::numeric_limits<int32_t>::max())
+            {
+                confirmations = spv::pspv->GetLastBlockHeight() - blockHeight + 1;
+            }
+
+            UniValue spentItem(UniValue::VOBJ);
+            spentItem.pushKV("txid", spent.ToString());
+            spentItem.pushKV("confirms", confirmations);
+            item.pushKV("spent", spentItem);
+        }
+        result.push_back(item);
+    }
+    return result;
+}
+
 std::string CSpvWrapper::GetRawTransactions(uint256& hash)
 {
     UInt256 spvHash;
     UIntConvert(hash.begin(), spvHash);
     return BRGetRawTransaction(wallet, spvHash);
 }
 
+std::string CSpvWrapper::GetHTLCSeed(uint8_t* md20)
+{
+    return BRGetHTLCSeed(wallet, md20);
+}
+
+uint64_t CSpvWrapper::GetFeeRate()
+{
+    return BRWalletFeePerKb(wallet);
+}
+
+HTLCDetails GetHTLCDetails(CScript& redeemScript)
+{
+    HTLCDetails script;
+
+    if (redeemScript[2] != 32)
```

Comment:
Can we verify that we have 32 bytes in redeemscript

Reply:
Now building up script length as we parse it and check it at each step.